src/python/pants/backend/rust/lint/rustfmt/rules_integration_test.py

```diff
@@ -0,0 +1,177 @@
+            RustfmtRequest.Batch(
+                "",
+                input_sources.snapshot.files,
+                partition_metadata=None,
+                snapshot=input_sources.snapshot
+            ),
+        ],
+    )
+    return fmt_result
+
+
+def get_snapshot(rule_runner: RuleRunner, source_files: dict[str, str]) -> Snapshot:
+    files = [FileContent(path, content.encode()) for path, content in source_files.items()]
+    digest = rule_runner.request(Digest, [CreateDigest(files)])
+    return rule_runner.request(Snapshot, [digest])
+
+
+def test_passing(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {"src/lib.rs": GOOD_FILE, "Cargo.toml": "", "BUILD": "rust_crate(name='crate')"}
+    )
+    tgt = rule_runner.get_target(Address("", target_name="crate"))
+    fmt_result = run_rustfmt(rule_runner, [tgt])
+
+    assert fmt_result.stdout.strip() == ""
+    assert fmt_result.output == get_snapshot(
+        rule_runner, {"src/lib.rs": GOOD_FILE, "Cargo.toml": ""}
+    )
+    assert fmt_result.did_change is False
+
+
+def test_failing(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {"src/lib.rs": BAD_FILE, "Cargo.toml": "", "BUILD": "rust_crate(name='crate')"}
+    )
+    tgt = rule_runner.get_target(Address("", target_name="crate"))
+    fmt_result = run_rustfmt(rule_runner, [tgt])
+    assert fmt_result.stderr == ""
+    assert fmt_result.output == get_snapshot(
+        rule_runner, {"src/lib.rs": FIXED_BAD_FILE, "Cargo.toml": ""}
+    )
+    assert fmt_result.did_change is True
+
+
+def test_mixed_sources(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "src/good.rs": GOOD_FILE,
+            "src/bad.rs": BAD_FILE,
+            "Cargo.toml": "",
+            "BUILD": "rust_crate(name='crate')",
+        }
+    )
+    tgt = rule_runner.get_target(Address("", target_name="crate"))
+    fmt_result = run_rustfmt(rule_runner, [tgt])
+    assert fmt_result.output == get_snapshot(
+        rule_runner, {"src/good.rs": GOOD_FILE, "src/bad.rs": FIXED_BAD_FILE, "Cargo.toml": ""}
+    )
+    assert fmt_result.did_change is True
+
+
+def test_multiple_targets(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "good/src/f.rs": GOOD_FILE,
+            "good/Cargo.toml": "",
+            "good/BUILD": "rust_crate()",
+            "bad/src/f.rs": BAD_FILE,
+            "bad/Cargo.toml": "",
+            "bad/BUILD": "rust_crate()",
+        }
+    )
+    tgts = [rule_runner.get_target(Address("good")), rule_runner.get_target(Address("bad"))]
+    fmt_result = run_rustfmt(rule_runner, tgts)
+    assert fmt_result.output == get_snapshot(
+        rule_runner,
+        {
+            "good/src/f.rs": GOOD_FILE,
+            "bad/src/f.rs": FIXED_BAD_FILE,
+            "good/Cargo.toml": "",
+            "bad/Cargo.toml": "",
+        },
+    )
+    assert fmt_result.did_change is True
+
+
+@pytest.mark.skip(reason="TODO")
```

Comment:
I am unsure as to why this would not work.

Reply:
What's the failure you're observing? rustfmt is still running?

I think there's two levels of 'skip':

1. global "never run rustfmt" via `--rustfmt-skip`, which sets `RustfmtSubsystem.skip`
2. local "don't run rustfmt on one crate" via `rust_crate(name="crate", skip_rustfmt=True)` in a `BUILD` file, which sets `SkipRustfmtField` for that crate

I think you've correctly implemented 2, but this test is for 1. So, there's potentially two steps to do to resolve this:

1. add support for the global `skip` flag. For example, what black does: https://github.com/pantsbuild/pants/blob/75b7be14501ff3228ef1eae4e8a413c371564913/src/python/pants/backend/python/lint/black/rules.py#L68-L73 (I note that this typically happens within an explicit partitioner, which this subsystem doesn't currently use. I'm not sure what the correct path forward there is, though.)
2. add a test for the per-crate `skip_rustfmt=True` flag

Reply:
Thanks! I will take two actions based on that comment, namely adding such flag check as well as adding a test for local skipping.

Reply:
@huonw I am actually a bit confused about a few things related to skip here.

So 1. Adding the flag in the build file does not seem to do the trick. Is the underlying check on the flag done as part of the engine?

2. I am confused about why we decide to have a skip flag replicated across each backend, and then we rely on such flag being tested against in each backend. Can we generalise it? I know the return type differs, but could it for example be done by ensuring that each return type has a skipped default value? (Please note that this point is more to get a throughout understanding, as it is completely unrelated to this PR and future work related to the PR)

3. If I am to make such check against the skip flag, would the proper return value be of type Partitions?

Reply:
I am also wondering why if it is really that simple as to add another flag to the sources target, why do no other tests exist that have tested such flag exist?

Reply:
Changes has occurred accordingly. It still does not fix the underlying issue wrt. skip. For now I consider it a problem unrelated to this PR unless someone can pinpoint me in a direction which indicates otherwise.

Reply:
So just to confirm, the file is or isn't being formatted?

Reply:
> I think about renaming FmtRequest to be more clear about inheritance.

@thejcannon: Maybe `AbstractFmtRequest` or similar name?

Reply:
@tdyas I will rename to that name in another PR

Reply:
> @tdyas I will rename to that name in another PR

Agreed; definitely out of scope for this PR. Also my comment was more directed toward @thejcannon for whenever they may do such a rename.

Reply:
Oops yeah my message had the wrong tense. Should've been "I'm thinking..."

Reply:
@thejcannon The test will remain skipped until I can see what the underlying issue is. 

The changes you specified about has been made accordingly.

Reply:
I don't feel comfortable merging this with a broken test, so let's keep digging on it

Reply:
I agree. Skip should work. I'll try this PR locally tonight to try and find out the issue. 

(My comment elsewhere in this PR about TODO comment is for the treatment of the sources field.)